src/Microsoft.NetCore.Analyzers/Core/Tasks/DoNotCreateTaskCompletionSourceWithWrongArguments.cs

```diff
@@ -0,0 +1,89 @@
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: null,
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // Ensure necessary types are available
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskCompletionSource, out var tcsType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemObject, out var objectType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskContinuationOptions, out var taskContinutationOptionsType))
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    // Ensure that you're using the TaskCompletionSource constructor with the object state property else exit
+                    var invocation = (IObjectCreationOperation)operationContext.Operation;
+                    if (!invocation.Type.OriginalDefinition.Equals(tcsType) ||
+                        invocation.Arguments.Length == 0 ||
+                        !invocation.Constructor.Parameters[0].Type.Equals(objectType))
+                    {
+                        return;
+                    }
+
+                    // Check to see if argument is a conversion operator
+                    if (!(invocation.Arguments[0] is IArgumentOperation arg) ||
+                        arg.Value.Kind != OperationKind.Conversion ||
+                        !(arg.Value is IConversionOperation conversionOperation))
+                    {
+                        return;
+                    }
+
+                    // Check to see if its an enum field or local
+                    if (conversionOperation.Operand.Kind != OperationKind.FieldReference &&
+                        conversionOperation.Operand.Kind != OperationKind.LocalReference)
```

Comment:
Will this generate false negatives for `OperationKind.ParameterReference` and `OperationKind.PropertyReference` passing in non-task continuation value? That is reasonable, but just confirming.

Reply:
Performing [PointsTo analysis](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Writing%20dataflow%20analysis%20based%20analyzers.md#well-known-flow-analyses) here can avoid these false negatives, but it might be an overkill.

Reply:
Good point on PropertyReference...I create a property in the test cases with the TaskContinuationOptions enum and passed that in.  Analyzer now handles that case.  But I'm a bit unsure how a ParameterReference works.  Could you provide me a simple code sample of what I should be looking for?